presto-main/src/main/java/com/facebook/presto/operator/aggregation/noisyaggregation/NoisyCountGaussianColumnRandomSeedAggregation.java

```diff
@@ -0,0 +1,161 @@
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, BIGINT.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
+                createInputParameterMetadata(type),
+                INPUT_FUNCTION,
+                COMBINE_FUNCTION,
+                OUTPUT_FUNCTION,
+                ImmutableList.of(new AccumulatorStateDescriptor(
+                        CountScaleRandomSeedState.class,
+                        stateSerializer,
+                        stateFactory)),
+                BIGINT);
+
+        Class<? extends Accumulator> accumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
+                Accumulator.class,
+                metadata,
+                classLoader);
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
+                GroupedAccumulator.class,
+                metadata,
+                classLoader);
+        return new BuiltInAggregationFunctionImplementation(NAME, inputTypes, ImmutableList.of(intermediateType), BIGINT,
+                true, false, metadata, accumulatorClass, groupedAccumulatorClass);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type type)
+    {
+        return ImmutableList.of(
+                new ParameterMetadata(STATE),
+                new ParameterMetadata(BLOCK_INPUT_CHANNEL, type),
+                new ParameterMetadata(BLOCK_INPUT_CHANNEL, DOUBLE),
+                new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, BIGINT),
+                new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void input(CountScaleRandomSeedState state, Block valueBlock, Block noiseScaleBlock, Block randomSeedBlock, int index)
+    {
+        double noiseScale = DOUBLE.getDouble(noiseScaleBlock, index);
+        if (noiseScale < 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Noise scale must be >= 0");
+        }
+        // Update count and retain scale and random seed
+        state.setCount(state.getCount() + 1);
+        state.setNoiseScale(noiseScale);
+        if (!randomSeedBlock.isNull(index)) {
+            state.setRandomSeed(BIGINT.getLong(randomSeedBlock, index));
+        }
+    }
+
+    public static void combine(CountScaleRandomSeedState state, CountScaleRandomSeedState otherState)
+    {
+        state.setCount(state.getCount() + otherState.getCount());
+        state.setNoiseScale(state.getNoiseScale() > 0 ? state.getNoiseScale() : otherState.getNoiseScale()); // noise scale should be > 0
+        state.setRandomSeed(state.getRandomSeed() > 0 ? state.getRandomSeed() : otherState.getRandomSeed());
+    }
+
+    public static void output(CountScaleRandomSeedState state, BlockBuilder out)
+    {
+        Random random = new SecureRandom();
+        if (state.getRandomSeed() > 0) {
```

Comment:
Unless I'm missing something, won't this aggregator only be used when the caller passes an argument for a random seed? And any `long` is a valid seed for `Random`, right? So maybe we should just use `new Random(state.getRandomSeed())` no matter what? If the caller wants to use `SecureRandom`, they should not pass a seed.

Reply:
We want to use secure random when seed <= 0 though. But now you say it, maybe just forget the <= 0 or > 0 part, and always use provided see with normal random. I will update the change